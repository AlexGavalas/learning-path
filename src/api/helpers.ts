import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
