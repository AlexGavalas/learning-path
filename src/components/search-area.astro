---
import { QUERY_FIELD_NAME } from '~constants';
import Loader from '~components/loader.astro';
import Button from '~components/button.astro';
import Input from './input.astro';

const { q } = Astro.props;
---

<search-area>
    <form class="relative flex h-16 items-center">
        <Input
            label="Search notes"
            name={QUERY_FIELD_NAME}
            autocomplete="off"
            placeholder="Type here"
            value={q}
        />
        <div class="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm">
            <Button>Search</Button>
        </div>
        <div id="loader" class="hidden absolute -bottom-1 h-[2px] w-full">
            <Loader />
        </div>
    </form>
</search-area>

<script>
    import { navigate } from 'astro:transitions/client';
    import { QUERY_FIELD_NAME } from '~constants';

    class SearchArea extends HTMLElement {
        formElement: HTMLFormElement | null = null;
        inputElement: HTMLInputElement | null = null;
        loaderElement: HTMLDivElement | null = null;

        constructor() {
            super();

            this.formElement = this.querySelector('form');
            this.inputElement = this.querySelector('input');
            this.loaderElement = this.querySelector('#loader');
        }

        handleKeyUp = (e: KeyboardEvent) => {
            if (e.key === '/') {
                this.inputElement?.focus();
            }
        };

        handleSearch = (e: SubmitEvent) => {
            this.loaderElement?.classList.remove('hidden');

            const url = new URL(location.href);
            const query = this.inputElement?.value ?? '';

            if (query.length > 0) {
                url.searchParams.set(QUERY_FIELD_NAME, query);
            } else {
                url.searchParams.delete(QUERY_FIELD_NAME);
            }

            // window.location.href = url.toString();
            navigate(url.toString());

            e.preventDefault();
        };

        connectedCallback() {
            document.addEventListener('keyup', this.handleKeyUp);
            this.formElement?.addEventListener('submit', this.handleSearch);
        }

        disconnectedCallback() {
            document.removeEventListener('keyup', this.handleKeyUp);
            this.formElement?.removeEventListener('submit', this.handleSearch);
        }
    }

    customElements.define('search-area', SearchArea);
</script>
