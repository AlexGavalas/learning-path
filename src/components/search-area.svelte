<script lang="ts">
    import { onDestroy, onMount } from 'svelte';

    import Button from '~components/button.svelte';
    import Input from '~components/input.svelte';
    import Loader from '~components/loader.svelte';
    import { QUERY_FIELD_NAME } from '~constants';

    interface $$Props {
        q: string;
    }

    export let q: $$Props['q'];

    let inputElement: any;
    let query = q;
    let loading = false;

    const keyPressHandler = (e: KeyboardEvent) => {
        if (e.key === '/') {
            e.stopImmediatePropagation();
            inputElement.focus();
        }
    };

    onMount(() => {
        typeof document !== 'undefined' &&
            document.addEventListener('keyup', keyPressHandler);
    });

    onDestroy(() => {
        typeof document !== 'undefined' &&
            document.removeEventListener('keyup', keyPressHandler);
    });

    const onSubmit = (e: SubmitEvent) => {
        e.preventDefault();

        loading = true;

        const url = new URL(location.href);

        if (query.length > 0) {
            url.searchParams.set(QUERY_FIELD_NAME, query);
        } else {
            url.searchParams.delete(QUERY_FIELD_NAME);
        }

        window.location.href = url.toString();
    };
</script>

<form on:submit={onSubmit} class="relative flex h-16 items-center">
    <Input
        bind:this={inputElement}
        bind:value={query}
        label="Search notes"
        name={QUERY_FIELD_NAME}
        autocomplete="off"
        placeholder="Type here"
    />
    <div class="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm">
        <Button>Search</Button>
    </div>
    {#if loading}
        <Loader />
    {/if}
</form>
