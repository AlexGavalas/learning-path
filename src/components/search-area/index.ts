export { SearchArea } from './search-area';
