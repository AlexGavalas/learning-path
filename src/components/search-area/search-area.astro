---
import { QUERY_FIELD_NAME } from '~constants';
import Loader from '~components/loader/linear.astro';
import Button from '~components/button/button.astro';
import Input from '~components/input/input.astro';
import cn from './search-area.module.css';

interface Props {
    query?: string;
}

const { query } = Astro.props;
---

<search-area>
    <form class={cn.container}>
        <Input
            autocomplete="off"
            class={cn.input}
            id={QUERY_FIELD_NAME}
            label="Search the notes"
            name={QUERY_FIELD_NAME}
            placeholder="Javascript, CSS grid ..."
            value={query}
        />
        <Button class={cn.button}>Search</Button>
        <div id="loader" class={cn.loader} hidden>
            <Loader />
        </div>
    </form>
</search-area>

<script>
    import { QUERY_FIELD_NAME } from '~constants';

    class SearchArea extends HTMLElement {
        formElement: HTMLFormElement | null = null;
        inputElement: HTMLInputElement | null = null;
        loaderElement: HTMLDivElement | null = null;

        constructor() {
            super();

            this.formElement = this.querySelector('form');
            this.inputElement = this.querySelector('input');
            this.loaderElement = this.querySelector('#loader');
        }

        handleKeyUp = (e: KeyboardEvent) => {
            if (e.key === '/') {
                this.inputElement?.focus();
            }
        };

        handleSearch = (e: SubmitEvent) => {
            if (this.loaderElement) {
                this.loaderElement.hidden = false;
            }

            const url = new URL(location.href);
            const query = this.inputElement?.value ?? '';

            if (query.length > 0) {
                url.searchParams.set(QUERY_FIELD_NAME, query);
            } else {
                url.searchParams.delete(QUERY_FIELD_NAME);
            }

            window.addEventListener(
                'pagehide',
                () => {
                    if (this.loaderElement) {
                        this.loaderElement.hidden = true;
                    }
                },
                { once: true },
            );

            window.location.href = url.toString();

            e.preventDefault();
        };

        connectedCallback() {
            const inputValue = this.inputElement?.value;

            if (inputValue) {
                this.inputElement?.setSelectionRange(
                    inputValue.length,
                    inputValue.length,
                );

                this.inputElement?.focus();
            }

            document.addEventListener('keyup', this.handleKeyUp);
            this.formElement?.addEventListener('submit', this.handleSearch);
        }

        disconnectedCallback() {
            document.removeEventListener('keyup', this.handleKeyUp);
            this.formElement?.removeEventListener('submit', this.handleSearch);
        }
    }

    customElements.define('search-area', SearchArea);
</script>
