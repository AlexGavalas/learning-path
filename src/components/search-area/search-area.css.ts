import { style } from '@vanilla-extract/css';

import { vars } from '../../styles/theme.css';

export const container = style({
    position: 'relative',
    display: 'grid',
    gap: vars.spacing.sm,
    gridTemplateColumns: '1fr auto',
    gridTemplateAreas: `"input ." "input button"`,
    '@media': {
        'screen and (max-width: 360px)': {
            gridTemplateColumns: '1fr',
        },
    },
});

export const input = style({
    display: 'grid',
    gap: vars.spacing.sm,
    gridArea: 'input',
    gridTemplateRows: 'subgrid',
});

export const searchButton = style({
    padding: `${vars.spacing.xs} ${vars.spacing.md} !important`,
    fontSize: '0.9em',
    gridArea: 'button',
});

export const loader = style({
    position: 'absolute',
    bottom: '-8px',
    height: '2px',
    width: '100%',
});
