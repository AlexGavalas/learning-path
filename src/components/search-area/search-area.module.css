.container {
    position: relative;
    display: grid;
    gap: var(--spacing-sm);
    grid-template-columns: 1fr auto;
    grid-template-areas: 'input .' 'input button';

    @media (max-width: 360px) {
        grid-template-columns: 1fr;
    }

    .input {
        display: grid;
        gap: var(--spacing-sm);
        grid-area: input;
        grid-template-rows: subgrid;
    }

    .button {
        padding: var(--spacing-xs) var(--spacing-md) !important;
        font-size: 0.9em;
        grid-area: button;
    }

    .loader {
        position: absolute;
        bottom: -8px;
        height: 2px;
        width: 100%;
    }
}
