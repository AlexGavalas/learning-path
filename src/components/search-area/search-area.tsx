import {
    type FormEventHandler,
    useCallback,
    useEffect,
    useRef,
    useState,
} from 'react';

import { Button } from '~components/button';
import { Input } from '~components/input';
import { Loader } from '~components/loader';
import { QUERY_FIELD_NAME } from '~constants';
import { useKeypress } from '~hooks/use-keypress';

export const SearchArea = ({ q }: { q: string }): JSX.Element => {
    const queryEl = useRef<HTMLInputElement | null>(null);

    const [query, setQuery] = useState(q);
    const [loading, setLoading] = useState(false);

    const keyPressHandler = useCallback((e: KeyboardEvent) => {
        e.stopImmediatePropagation();
        queryEl.current?.focus();
    }, []);

    useEffect(() => {
        setLoading(false);
    }, [q]);

    useKeypress('/', keyPressHandler);

    const onSubmit: FormEventHandler<HTMLFormElement> = (e) => {
        e.preventDefault();

        setLoading(true);

        const url = new URL(location.href);

        if (query.length > 0) {
            url.searchParams.set(QUERY_FIELD_NAME, query);
        } else {
            url.searchParams.delete(QUERY_FIELD_NAME);
        }

        window.location.href = url.toString();
    };

    return (
        <form onSubmit={onSubmit} className="relative flex h-16 items-center">
            <Input
                label="Search notes"
                name={QUERY_FIELD_NAME}
                ref={queryEl}
                value={query}
                onChange={(e) => {
                    setQuery(e.target.value);
                }}
                autoFocus={query.length > 0}
                autoComplete="off"
                placeholder="Type here"
            />
            <div className="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm">
                <Button>Search</Button>
            </div>
            {loading && <Loader />}
        </form>
    );
};
