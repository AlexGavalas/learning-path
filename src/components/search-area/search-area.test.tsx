import { axe } from 'jest-axe';

import { renderWithUser, screen } from '~test/helpers';

import { SearchArea } from './search-area';

const baseURL = new URL('http://localhost/');

describe('<SearchArea />', () => {
    beforeAll(() => {
        jest.spyOn(window, 'location', 'get').mockReturnValue({
            ancestorOrigins: [] as unknown as DOMStringList,
            hash: '',
            host: baseURL.host,
            hostname: baseURL.hostname,
            href: baseURL.href,
            origin: baseURL.origin,
            pathname: baseURL.pathname,
            port: baseURL.port,
            protocol: baseURL.protocol,
            search: baseURL.search,
            assign: jest.fn(),
            reload: jest.fn(),
            replace: jest.fn(),
            toString: jest.fn(),
        });
    });

    it('renders', () => {
        const { container } = renderWithUser(<SearchArea q="" />);

        expect(container).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const { container } = renderWithUser(<SearchArea q="" />);

        const a11yResults = await axe(container);

        expect(a11yResults).toHaveNoViolations();
    });

    describe('when user presses "/"', () => {
        it('search input receives focus', async () => {
            const { user } = renderWithUser(<SearchArea q="" />);

            expect(screen.getByRole('textbox')).not.toHaveFocus();

            await user.keyboard('/');

            expect(screen.getByRole('textbox')).toHaveFocus();
        });
    });

    describe('when user types in search input', () => {
        describe('when user clicks search', () => {
            it('does nothing when the search query is empty', async () => {
                const { user } = renderWithUser(<SearchArea q="" />);

                await user.click(
                    screen.getByRole('button', { name: /search/i }),
                );

                const url = new URL(baseURL);
                url.searchParams.delete('q');

                // TODO
            });

            it('navigates the user when it has a query', async () => {
                const { user } = renderWithUser(<SearchArea q="" />);
                const searchQuery = 'abc';

                await user.keyboard(`/${searchQuery}`);

                await user.click(
                    screen.getByRole('button', { name: /search/i }),
                );

                const url = new URL(baseURL);
                url.searchParams.set('q', searchQuery);

                // TODO
            });
        });
    });
});
