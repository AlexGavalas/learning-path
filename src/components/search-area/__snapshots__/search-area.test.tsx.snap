// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchArea /> renders 1`] = `
<div>
  <form
    class="relative flex h-16 items-center"
  >
    <label
      class="flex h-full w-full flex-col"
    >
      <span>
        Search notes
      </span>
      <input
        autocomplete="off"
        class="h-1/2 rounded border-[1px] border-solid border-light-border bg-transparent pl-2 text-sm dark:border-0  dark:bg-neutral-800 dark:placeholder-gray-300 dark:placeholder-opacity-60"
        name="q"
        placeholder="Type here"
        value=""
      />
    </label>
    <div
      class="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm"
    >
      <button
        class="flex items-center rounded border-0 p-2 bg-slate-300 hover:bg-slate-300/75 dark:bg-slate-600 dark:hover:bg-slate-600/75"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;
